### DISCUSSION TITLE
Setting up the Grammarly LSP

### DISCUSSION THREAD

**Discussion by JurajLieskovsky** [helix-editor/helix#7383] [Q&A]
[multiple language server support](https://github.com/helix-editor/helix/pull/2507) being recently pulled into the master branch I would like to set up a Grammarly LSP, mainly for checking my grammar in markdown files.

So far I have installed the LSP using the recommended way i.e. 
```
git clone git@github.com:znck/grammarly.git
```
and then
```
pnpm install
pnpm run build
```
having to install a specific version of Node.js with [nvm](https://github.com/nvm-sh/nvm) as otherwise (including installing the lsp just with npm) it would not pass `test`s.

Then I added some initial lines to `.config/helix/languages.toml`
```
[language-server.grammarly]
command = "/opt/grammarly/extension/node_modules/.bin/grammarly-languageserver"
args = ["--stdio"]
config.clientId = "client_BaDkMgx4X19X9UxxYRCXZo"

[[language]]
name = "markdown"
language-servers = ["marksman", "grammarly"]
```

With this configuration, I am getting the error message:
```
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "node:internal/deps/undici/undici:11522\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "    Error.captureStackTrace(err, this);\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "          ^\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "TypeError: Failed to parse URL from /opt/grammarly/node_modules/.pnpm/web-tree-sitter@0.20.5/node_modules/web-tree-sitter/tree-sitter.wasm\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "    at Object.fetch (node:internal/deps/undici/undici:11522:11) {\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "  [cause]: TypeError [ERR_INVALID_URL]: Invalid URL\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "      at new NodeError (node:internal/errors:399:5)\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "      at new URL (node:internal/url:743:13)\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "      at new Request (node:internal/deps/undici/undici:7082:25)\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "      at fetch2 (node:internal/deps/undici/undici:10661:25)\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "      at Object.fetch (node:internal/deps/undici/undici:11520:18)\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "      at fetch (node:internal/process/pre_execution:239:25)\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "      at /opt/grammarly/node_modules/.pnpm/web-tree-sitter@0.20.5/node_modules/web-tree-sitter/tree-sitter.js:1:15192\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "      at /opt/grammarly/node_modules/.pnpm/web-tree-sitter@0.20.5/node_modules/web-tree-sitter/tree-sitter.js:1:15413\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "      at new Promise (<anonymous>)\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "      at Parser.init (/opt/grammarly/node_modules/.pnpm/web-tree-sitter@0.20.5/node_modules/web-tree-sitter/tree-sitter.js:1:332) {\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "    input: '/opt/grammarly/node_modules/.pnpm/web-tree-sitter@0.20.5/node_modules/web-tree-sitter/tree-sitter.wasm',\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "    code: 'ERR_INVALID_URL'\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "  }\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "}\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "\n"
2023-06-19T09:02:46.717 helix_lsp::transport [ERROR] grammarly err <- "Node.js v20.2.0\n"
2023-06-19T09:02:46.721 helix_lsp::transport [ERROR] grammarly err: <- StreamClosed
```
at startup. After that I get
```
2023-06-19T09:19:42.444 helix_lsp::transport [ERROR] grammarly err: <- IO(Os { code: 32, kind: BrokenPipe, message: "Broken pipe" })
```
over and over.

There are likely multiple problems one of which might be how I add the `ClientId`. In the discussion [Need help configuring the language server with neovim](https://github.com/znck/grammarly/discussions/285), the `ClientId` is added through `init_options` while in sublime-text it is under `initializationOptions` inside the config. I looked through helix's default [`languages.toml`](https://github.com/helix-editor/helix/blob/master/languages.toml) but found no similar key in any of the configs.

So in conclusion, any help would be greatly appreciated. I'll try to figure out what's wrong in my spare time but I'm in over my head.

**Comment by pascalkuthe** [comment]
https://github.com/znck/grammarly/discussions/285#discussioncomment-4441853

Seems like you need to run the test commands to build some TS grammar files

**Reply by JurajLieskovsky** [reply]
I ran the `pnpm test` command, do you mean something different? In the readme linked to the comment, I see no further mention of tests.

**Reply by pascalkuthe** [reply]
It sounds like that exact sequence of commands they descrived is needed (I guess that the tests build the TS grammars and reinstalling them fixes the problem). Inoly skimmed that discussiom tough and he seems to habe run into the same issue that's why I linked it. I don't use the grammarly LS myself

**Reply by JurajLieskovsky** [reply]
Ok, now I am getting
```
2023-06-19T10:57:31.098 helix_lsp::transport [ERROR] grammarly <- InternalError: Request initialize failed with message: fetch failed
2023-06-19T10:57:31.098 helix_lsp [ERROR] failed to initialize language server: protocol error: InternalError: Request initialize failed with message: fetch failed
```

It didn't occur to me that running the sequence of commands twice would produce a different result.

**Comment by pascalkuthe** [comment]
Btw. I would recommend ltex-ls since it's opensource/fully local (no calls to external API) and doesn't harvest your data

**Reply by JurajLieskovsky** [reply]
Huh, looks like something I would be interested in trying. Did you set it up for yourself in helix?

**Reply by pascalkuthe** [reply]
Yeah the config is really straight forward

**Reply by JurajLieskovsky** [reply]
I can confirm that it really is straightforward, thank you for the suggestion.

**Reply by tapyu** [reply]
Hi! In my case, I am trying to use a grammar LSP for typesetting systems, such as LaTeX and Typist. Unfortunately, this extension does seem to support the latter one :( do you have any alternative? In Vs code, I've been using the Grammarly extension for both and it is working well.

**Reply by tapyu** [reply]
For those who are trying to find a powerful grammar checker: unfortunately, ltex-ls is definitely not as powerful as Grammarly. `ltex-ls` seemed great for basic stuffs, but only Grammarly can check more complex English structures. For me, I still rely on Grammarly.
